### DISCUSSION TITLE
Migrate Groups and Users?

### DISCUSSION THREAD

**Discussion by tclemans** [outline/outline#4823] [Self Hosting] [answered]
What is the best way to migrate from one Outline self-hosted instance to another?

I am planning to migrate from a docker-compose self-hosted instance to a Kubernetes self-hosted instance. I know I can export/import the Documents and Images easily enough, but how can I move over Groups and Users? 

Login is via OIDC and the new instance will be the same as the old in every way. 

I might just have to manually recreate the Groups, etc. But I'm hoping there's a better way. 

Thanks!

**Comment by tommoor** [comment]
The best way in this case is just to transfer the entire database using pg_dump or similar. Any reason you wouldn’t do this?

**Reply by tclemans** [reply]
I should be able to do that. I'll be moving from S3 to a self hosted MinIO instance well, so I think I'll have to sync those too. Unless that info is in the pg_dump?

**Reply by tommoor** [reply]
It's not, you'd need to copy the entire bucket.

**Reply by tommoor** [reply]
If you update both installations to the latest commit there is a new export/import as JSON option – this would work too but it's lossy, not everything is transferred.

**Reply by tclemans** [reply]
I'll try pg_dump and copying data from S3 to MinIO. Thanks!

**Comment by tclemans** [comment, marked as answer]
I successfully lifted the database and uploads from one Outline instance to another. 

## Solution:


**Move Docs and Settings**
1. Export Database with `pg_dump --clean --if-exists dbname > dumpfile`
2. Move dumpfile to new instance and import with `psql dbname < dumpfile` 
    a. Source and Destination database should have same username and database name. 
3. The SECRET_KEY needs to be the same for both Source and Destination (unless you figure out how to update it properly, i.e. #1441)

**Copy Uploads from S3 to MinIO**
1. Install `mc` via [mc docs](https://min.io/docs/minio/linux/reference/minio-mc.html)
2. Setup Aliases for S3 and MinIO buckets
    a.  `mc alias set minio https://minio.mydomain.com username password`
    b.  `mc alias set s3 https://s3.amazonaws.com username password`
3. Sync buckets with `mc mirror s3/bucket-name minio/new-bucket-name`
4. Make sure Outline env is set to: `AWS_S3_FORCE_PATH_STYLE: "true"` (you may need to restart container/pod for this to take effect).

If you get `413` Errors when running `mc mirror` you need to adjust nginx ingress size checking. See [Set up Nginx proxy with MinIO Server](https://min.io/docs/minio/linux/integrations/setup-nginx-proxy-with-minio.html)
